activesupport/lib/active_support/cache/streaming_compressor.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+require "zlib"
+require "concurrent/executor/cached_thread_pool"
+require "active_support/core_ext/numeric/bytes"
+
+module ActiveSupport
+  module Cache
+    # Serializes object and compresses serialized representation in streaming way.
+    # This avoids storing whole copy of serialized intermediate representation,
+    # like this happens in `Zlib.deflate(Marshal.dump(value))`.
+    module StreamingCompressor
+      DEFLATE_BUFFER_SIZE = 16.kilobytes.to_i
+      INFLATE_CHUNK_SIZE = 4.kilobyte.to_i
+
+      extend self
+
+      # Dumps value into string. Returns nil if `:compress_threshold` option
+      # is set and serialized value is less than given value.
+      def dump(value, **options)
+        io = StringIO.new
+        raw_bytes = dump_to_io(value, io, **options)
+        return unless raw_bytes
+        io.rewind
+        result = io.read
+        result if result.bytesize < raw_bytes
+      end
+
+      # Same as dump but streams compressed data. Returns nil and don't stream
+      # anything if `:compress_threshold` option is set
+      # and serialized value is less than given value.
+      def dump_to_io(value, to_io, compress_threshold: nil)
+        rio, wio = binary_pipe
+        within_safe_thread do
+          begin
+            Marshal.dump(value, wio)
```

Comment:
`wio.write(value)` should be faster than `Marshal.dump(value, wio)`.
https://gist.github.com/ohaddahan/6946f93ae4f867d663529ed8fe4ace25

Reply:
This is key part of streaming compressor. Marshal.dump(value) creates string with whole serialized value. While Marshal.dump(value, io) dumps it to stream and paused, when stream is full, also discarding data which have already been read. Stream buffer size is defined by OS.

Reply:
You're opening the `IO` on `#33` and ensuring closer on `#38`.
I don't see any logic change between `Marshal.dump(value, io)` vs. `wio.write(Mashsal.dump(value))`.
Can you please explain?

Reply:
@printercu This was confusing to me as well, `Marshal.dump` doesn't actually return a string object, but rather an IO object. I would still benchmark before making any kind of recommendation to change.

@ohaddahan you can benchmark this using the provided code in this comment (collapsed) https://github.com/rails/rails/pull/33920#issuecomment-423125837.

You can make a new rails example app from master https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#running-an-application-against-your-local-branch Then change it to point to a local copy or rails using `path:` in the gemfile. 

You can then get these changes in your local copy using 

```
git checkout -b printercu-streaming_compressor master
git pull https://github.com/printercu/rails.git streaming_compressor
```

From there you can make changes and benchmark yourself.  Actually looks like the branch is over here https://github.com/printercu/rails/pull/1/files.

Reply:
@schneems @printercu not seeing a noticeable gain on medium + sizes.
https://gist.github.com/ohaddahan/ccf2abf42ef3034b13b3a5ee277ef0e0

BUT , when I tried benchmarking `wio.sync` I got consistent crashes. I'm not 100% sure this is thread safe.

Reply:
Looks like there is mistake in rubydoc:
```
[2] pry(main)> Marshal.dump(1).class
=> String < Object
[3] pry(main)> Marshal.dump(1, StringIO.new).class
=> StringIO < Data
```

Reply:
@printercu you're correct, confirmed here.
https://gist.github.com/ohaddahan/c1bd9ec897b8675994f25a2101728e4e

Reply:
`Marshal.dump` returns the `IO` it received.
`x = StringIO.new
y =  Marshal.dump(1, x)
x.object_id == y.object_id`